### PR TITLE
[bugfix] fix label effects

### PR DESCRIPTION
**Pull request by PatriceJiang**
ref https://github.com/cocos2d/cocos2d-x/issues/19551

- shadow/outline 
- batch command

### CODE REVIEW

File: cocos/2d/CCLabel.cpp [hunk 1]
```diff
@@ -1705,48 +1781,61 @@ void Label::draw(Renderer *renderer, const Mat4 &transform, uint32_t flags)
...
         }
     }
 }
 
 void Label::updateBlendState()
 {
-    backend::BlendDescriptor& blendDescriptor = _customCommand.getPipelineDescriptor().blendDescriptor;
-    blendDescriptor.blendEnabled = true;
-    if (_blendFunc == BlendFunc::ALPHA_NON_PREMULTIPLIED)
+    setOpacityModifyRGB(_blendFunc == BlendFunc::ALPHA_NON_PREMULTIPLIED);
```
Comment (minggo) [inline cocos/2d/CCLabel.cpp]: it seems the logic is different from previous's logic

File: cocos/2d/CCLabel.h [hunk 2]
```diff
@@ -749,9 +762,8 @@ class CC_DLL Label : public Node, public LabelProtocol, public BlendProtocol
     Color4F _textColorF;
 
     QuadCommand _quadCommand;
-    CustomCommand _customCommand;
-    CustomCommand _customCommandOutLine;
-    CustomCommand _customCommandShadow;
+
+    std::vector<BatchCommand> _batchCommands;
```
Comment (minggo) [inline cocos/2d/CCLabel.h:766]: Why need vector?

File: cocos/2d/CCFontAtlas.h [hunk 3]
```diff
@@ -139,6 +141,7 @@ class CC_DLL FontAtlas : public Ref
     void* _iconv;
 
     // Dynamic GlyphCollection related stuff
+    Texture2D::PixelFormat _currentPixelFormat;
```
Comment (minggo) [inline cocos/2d/CCFontAtlas.h]: It seems `_currentPixelFormat ` is not needed.

File: cocos/2d/CCFontAtlas.cpp [hunk 4]
```diff
@@ -153,6 +146,16 @@ FontAtlas::~FontAtlas()
...
 
+void FontAtlas::initTextureWithZeros(Texture2D *texture)
+{
+    char *zeros = nullptr;
+    size_t cnt = _currentPixelFormat == Texture2D::PixelFormat::RGBA8888 ? CacheTextureWidth * CacheTextureWidth * 4 : CacheTextureWidth * CacheTextureWidth;
+    zeros = new char[cnt];
+    std::fill(zeros, zeros + cnt, 0);
+    texture->initWithData(zeros, cnt, _currentPixelFormat, CacheTextureWidth, CacheTextureHeight, Size(CacheTextureWidth, CacheTextureHeight));
+    delete[] zeros;
+}
```
Comment (minggo) [inline cocos/2d/CCFontAtlas.cpp:167]: I think this function make codes not clear as before.

File: cocos/2d/CCLabel.h (hunk 2 above)
Comment (Mee-gu) [inline cocos/2d/CCLabel.h:766]: _batchCommands does not seem to be set?

File: cocos/2d/CCLabel.cpp (hunk 1 above)
Comment (PatriceJiang) [inline cocos/2d/CCLabel.cpp]:
```suggestion
    setOpacityModifyRGB(_blendFunc != BlendFunc::ALPHA_NON_PREMULTIPLIED);
```

File: cocos/2d/CCLabel.h (hunk 2 above)
Comment (PatriceJiang) [reply cocos/2d/CCLabel.h:766]: Each `_batchNodes` need a group of different renderer commands.

Comment (PatriceJiang) [reply cocos/2d/CCLabel.h:766]: It is set in `Label::draw` though `updateShaderProgram`
```cpp
            if (_batchCommands.size() != _batchNodes.size())
            {
                _batchCommands.resize(_batchNodes.size());
                updateShaderProgram();
            }
```

File: cocos/2d/CCFontAtlas.cpp (hunk 4 above)
Comment (PatriceJiang) [reply cocos/2d/CCFontAtlas.cpp:167]: This procedure is also invoked in `prepareLetterDefinitions()`.
Is it a naming issue?

File: cocos/2d/CCFontAtlas.h (hunk 3 above)
Comment (PatriceJiang) [reply cocos/2d/CCFontAtlas.h]: Yes, I will remove it

File: cocos/2d/CCFontAtlas.cpp [hunk 5]
```diff
@@ -413,9 +419,8 @@ bool FontAtlas::prepareLetterDefinitions(const std::u32string& utf32Text)
                     {
                         tex->setAliasTexParameters();
                     }
-                    tex->initWithData(_currentPageData, _currentPageDataSize,
-                        pixelFormat, CacheTextureWidth, CacheTextureHeight, Size(CacheTextureWidth, CacheTextureHeight));
```
Comment (PatriceJiang) [inline cocos/2d/CCFontAtlas.cpp:417]: The underline texture format should be `RGBA8888` or `A8`,  here the `pixelFormat` is `AI88` or `A8`.

Comment (Mee-gu) [reply cocos/2d/CCFontAtlas.cpp:417]: AI88 does not support in metal, so I think should change it to RGBA8888

Comment (PatriceJiang) [reply cocos/2d/CCFontAtlas.cpp:417]: yes, the format is mapped in `initTextureWithZeros`

File: cocos/2d/CCLabel.h [hunk 6]
```diff
@@ -804,7 +816,7 @@ class CC_DLL Label : public Node, public LabelProtocol, public BlendProtocol
     backend::UniformLocation _effectColorLocation;
     backend::UniformLocation _effectTypeLocation;
     
-    backend::ProgramState* _programState = nullptr;
+    backend::ProgramState* _programState        = nullptr;
```
Comment (ilqvya) [inline cocos/2d/CCLabel.h]: ?

**Pull request merged**